<a href="https://colab.research.google.com/github/ShaazCode/ML-Projects-Medical/blob/main/Heart_Diseases_Prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Heart Disease Prediction Project using Logistic Regression**

Here we're going to build an heart diseases prediction model using logistic regreeesion

We're going to do binary classification weather a person is having heart disease or not (Yes/No)


*Workflow*
Step 1: Collection of data

Step 2: Data Pre-processing

Step 3: Train/Test splits

Step 4: Train the model (logistic regression)

Step 5: Evaluation


**Importing libraries (dependencies)**

In [1]:
import pandas as pd
import numpy as np

**Column details**

1. Age : Its contain the age of person

2. Sex: It contain the gender of person

3. CP(Chest pain) : It contain the chest pain data.

4. Resting blood pressure : (Resting blood pressure means the flow of your blood against artery walls when you're calm

5. Serum chorestol in. mg/dl

6. Fasting blood sugar

7. Resting electrocardiographics results (1,2,3)

8.  Maximun heart rates achieved

9. Exercise included angina

10. oldpeak = ST depression included by exercise relative to rest

11. The slope of the peak exercise ST segment

12. Number of measures vessels

13. thal 3 = normal; 6 = fixed detected ; 7 = Reversable detect

If we're getting the output as 0 then person is not having heart diseases and if we will get any other values then the person is having heart diseases




### **Data collection**

In [ ]:
heart_diseases = pd.read_csv('/content/heart_disease_data.csv')

In [4]:
heart_diseases

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63    1   3       145   233    1        0      150      0      2.3   
1     37    1   2       130   250    0        1      187      0      3.5   
2     41    0   1       130   204    0        0      172      0      1.4   
3     56    1   1       120   236    0        1      178      0      0.8   
4     57    0   0       120   354    0        1      163      1      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope  ca  thal  target  
0        0   0     1       1  
1        0   0     2       1  
2        2   0     2       1  
3        2   0     2       1  
4        2   0     2       1  
..     ...  ..   ...     ...  
298      1   0     3       0  
299      1   0     3       0  
300      1   2     3       0  
301      1   1     3       0  
302      1   1     2       0  

[303 rows x 14 columns]

**Data Pre-processing (EDA) Exploratatory data analysis**

In [5]:
# Printing first 5 rows

heart_diseases.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

In [6]:
heart_diseases.head(10)

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   
5   57    1   0       140   192    0        1      148      0      0.4      1   
6   56    0   1       140   294    0        0      153      0      1.3      1   
7   44    1   1       120   263    0        1      173      0      0.0      2   
8   52    1   2       172   199    1        1      162      0      0.5      2   
9   57    1   2       150   168    0        1      174      0      1.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1  
5   0     1       1  
6   0     2       1  
7   0     3       1  
8   0     3       1  
9   0     2       1

In [7]:
# Printing last 5 values from datasets

In [8]:
heart_diseases.tail()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope  ca  thal  target  
298      1   0     3       0  
299      1   0     3       0  
300      1   2     3       0  
301      1   1     3       0  
302      1   1     2       0

In [9]:
heart_diseases.tail(10)

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
293   67    1   2       152   212    0        0      150      0      0.8   
294   44    1   0       120   169    0        1      144      1      2.8   
295   63    1   0       140   187    0        0      144      1      4.0   
296   63    0   0       124   197    0        1      136      1      0.0   
297   59    1   0       164   176    1        0       90      0      1.0   
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope  ca  thal  target  
293      1   0     3       0  
294      0   0     1       0  
295      2   2     3       0  
296      1   0     2       0  
297      1   2     1       0  
298      1   0     3       0  
299      1   0     3       0  
300      1   2     3       0  
301      1   1     3       0  
302      1   1     2       0

In [10]:
# get the number of rows and column from dataset

heart_diseases.shape

(303, 14)

In [11]:
# get information of dataset

heart_diseases.info()

# here we got the information of data set like which column is containing how much values what is overall count
# what is the data type of column, etc

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        303 non-null    int64  
 12  thal      303 non-null    int64  
 13  target    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


In [13]:
heart_diseases.isnull().sum()

# here there's no any missing values, already processed data

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0
dtype: int64

In [14]:
# statistical of data
heart_diseases.describe()

# this describe only work on number column only , here we are getting all stats of all column

age         sex          cp    trestbps        chol         fbs  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean    54.366337    0.683168    0.966997  131.623762  246.264026    0.148515   
std      9.082101    0.466011    1.032052   17.538143   51.830751    0.356198   
min     29.000000    0.000000    0.000000   94.000000  126.000000    0.000000   
25%     47.500000    0.000000    0.000000  120.000000  211.000000    0.000000   
50%     55.000000    1.000000    1.000000  130.000000  240.000000    0.000000   
75%     61.000000    1.000000    2.000000  140.000000  274.500000    0.000000   
max     77.000000    1.000000    3.000000  200.000000  564.000000    1.000000   

          restecg     thalach       exang     oldpeak       slope          ca  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean     0.528053  149.646865    0.326733    1.039604    1.399340    0.729373   
std      0.525860   22.905161    0.469794    1.161075    0.616226    1.022606   
min      0.000000   71.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000  133.500000    0.000000    0.000000    1.000000    0.000000   
50%      1.000000  153.000000    0.000000    0.800000    1.000000    0.000000   
75%      1.000000  166.000000    1.000000    1.600000    2.000000    1.000000   
max      2.000000  202.000000    1.000000    6.200000    2.000000    4.000000   

             thal      target  
count  303.000000  303.000000  
mean     2.313531    0.544554  
std      0.612277    0.498835  
min      0.000000    0.000000  
25%      2.000000    0.000000  
50%      2.000000    1.000000  
75%      3.000000    1.000000  
max      3.000000    1.000000

In [15]:
# checking the distribution of target variable

heart_diseases['target'].value_counts()

# so here we're having an column of target overall we're having an 303 values from that
# 165 people are having diseases and 138 people are not having diseases

target
1    165
0    138
Name: count, dtype: int64

**Splitting the features and target column**

In [16]:
# So as we know in our dataset Target is the output column so we will sepearte it.

In [17]:
X = heart_diseases.drop(columns='target', axis=1)

Y = heart_diseases['target']

In [22]:
print(X.to_string())

     age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  ca  thal
0     63    1   3       145   233    1        0      150      0      2.3      0   0     1
1     37    1   2       130   250    0        1      187      0      3.5      0   0     2
2     41    0   1       130   204    0        0      172      0      1.4      2   0     2
3     56    1   1       120   236    0        1      178      0      0.8      2   0     2
4     57    0   0       120   354    0        1      163      1      0.6      2   0     2
5     57    1   0       140   192    0        1      148      0      0.4      1   0     1
6     56    0   1       140   294    0        0      153      0      1.3      1   0     2
7     44    1   1       120   263    0        1      173      0      0.0      2   0     3
8     52    1   2       172   199    1        1      162      0      0.5      2   0     3
9     57    1   2       150   168    0        1      174      0      1.6      2   0     2
10    54  

In [23]:
# above we had deleted the target column successfully

In [24]:
print(Y)
# here we can see there's a column of target which we're storing inside the Y

0      1
1      1
2      1
3      1
4      1
      ..
298    0
299    0
300    0
301    0
302    0
Name: target, Length: 303, dtype: int64


**Training & Testing the data **

In [27]:
from sklearn.model_selection import train_test_split

In [28]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, stratify=Y, random_state=2)

In [ ]:
# Test_size = 0.2 => 20% for testing and 80% for training
# Stratify = this help us to prevent with imbalanced data
# Random_state = you get the same split every split at samne time if you wont say it then every split will differently

In [31]:
print(X.shape, X_train.shape, X_test.shape)

(303, 13) (242, 13) (61, 13)


In [32]:
# so here we're gettting to know that x is having total 303 values then x train is containing 242 values and x test is containing 61 values

**Model Training**

In [34]:
from sklearn.linear_model import LogisticRegression

In [38]:
# logistic Regression model

model = LogisticRegression()

In [36]:
# Training the Logistic regression model

# we can make use of train.fit to train the logisistic model

In [39]:
model.fit(X_train, Y_train)

/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [41]:
# above we had trained the model quickly becuase the dataset is very low

**Model Evaluation**

In [43]:
from sklearn.metrics import accuracy_score

In [44]:
# Accuracy of training data

In [45]:
X_train_prediction = model.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction, Y_train)

In [46]:
print("Accuracy on training data :", training_data_accuracy)

Accuracy on training data : 0.8512396694214877


So here we're getting the accuracy as 0.8512 it means the accuracy is 85%
somehow the 85% accuracy is also good but we have to boost it

In [49]:
# Accuracy for test data

X_test_prediction = model.predict(X_test)
test_data_accuracy = accuracy_score(X_test_prediction, Y_test)

In [50]:
print("Accuracy of test data:", test_data_accuracy)

Accuracy of test data: 0.819672131147541


So here the accuracy on test data is 81%

In [ ]:
# Training accuracy means => How the model trained from the given data

# Testing accuracy means => How well the model perform on new data unseen data

Building a predictive model

In [51]:
input_data = (61,1,2,150,243,1,1,137,1,1.0,1,0,2)

# here the input data is into tuple convert it into array

input_data_as_numpy_array = np.asarray(input_data)


# reshape the numpy array as we are predicting

input_data_reshaped = input_data_as_numpy_array.reshape(1, -1)

prediction = model.predict(input_data_reshaped)

print(prediction)

[1]


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(


In [52]:
if(prediction[0] == 0):
  print("The person does not have heart disease")
else:
  print("The person have heart diseases")

The person have heart diseases


**Code for pushing on github**

In [56]:
from google.colab import auth
auth.authenticate_user()

# conecting google collab

In [57]:
!git config --global user.name "ShaazCode"


In [58]:
!git config --global user.email "shaaz.code@gmail.com"


In [59]:
!git clone https://github.com/ShaazCode/ML-Projects-Medical.git


Cloning into 'ML-Projects-Medical'...


In [65]:
cd /content/ML-Projects-Medical

/content/ML-Projects-Medical


In [66]:
cd ML-Projects-Medical

[Errno 2] No such file or directory: 'ML-Projects-Medical'
/content/ML-Projects-Medical
